mindmeld/models/nn_utils/classification.py

```diff
@@ -0,0 +1,559 @@
+except ImportError:
+    nn_module = object
+    pass
+
+SEED = 6174
+LABEL_PAD_TOKEN_IDX = -1  # value set based on default label padding idx in pytorch
+
+logger = logging.getLogger(__name__)
+
+
+class BaseClassification(nn_module):
+    """
+    A base class for sequence & token classification using deep neural nets. Both the classification
+     submodules have a common fit() method defined in this base class, which also drives the
+     training of pytorch based deep nets. The net's computational graph is defined only when the
+     fit() method is called. This base class also holds few common utility methods and
+     further defines the skeleton of the children classes through abstract methods.
+    """
+
+    def __init__(self):
+        super().__init__()
+
+        self.name = self.__class__.__name__
+        self.params = Bunch()
+        self.params["name"] = self.name
+        self.encoder = None
+
+        self.ready = False  # True when .fit() is called or loaded from a checkpoint, else False
+        self.dirty = False  # True when the model weights aren't saved to disk yet, else False
+
+        self.out_dim = float('-inf')
+
+    def __repr__(self):
+        return f"<{self.name}> ready:{self.ready} dirty:{self.dirty}"
+
+    def get_default_params(self) -> Dict:
+        return get_default_params(self.__class__.__name__)
+
+    def log_and_return_model_info(self, verbose: bool = False) -> str:
+        """
+        Logs and returns the details of the underlying torch.nn model, such as occupying disk space
+        when dumped, number of parameters, the device on which the model is placed, etc.
+
+        Args:
+            verbose (bool): Determines the amount of information to be logged and returned.
+        """
+        msg = f"{self.name} " \
+              f"ready:{self.ready} dirty:{self.dirty} device:{self.params.device} " \
+              f"\n\tNumber of weights (trainable, all):{get_num_weights_of_model(self)} "
+        verbose_msg = msg + (
+            f"\n\tDisk Size (in MB): {get_disk_space_of_model(self):.4f} " if verbose else ""
+        )
+        logger.info(verbose_msg)
+        return msg
+
+    def to_device(self, batch_data: BatchData) -> BatchData:
+        """
+        Places pytorch tensors on the device configured through the params
+
+        Args:
+            batch_data (BatchData): A BatchData object consisting different tensor objects
+        """
+        for k, v in batch_data.items():
+            if v is not None and isinstance(v, torch.Tensor):
+                batch_data[k] = v.to(self.params.device)
+            elif isinstance(v, list):
+                batch_data[k] = [
+                    vv.to(self.params.device) if isinstance(vv, torch.Tensor) else vv
+                    for vv in v
+                ]
+            elif isinstance(v, dict):
+                batch_data[k] = self.to_device(batch_data[k])
+        return batch_data
+
+    # pylint: disable=too-many-locals
```

Comment:
Would it help to break this function down into a helper and main. The main loop can go in a ``_fit`` to contain the local variables and easier readability.